src/protocols/secure_channel/SessionEstablishmentTransport.h

```diff
@@ -0,0 +1,72 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines Application Channel class. The object of this
+ *      class can be used by CHIP data model cluster applications to send
+ *      and receive messages. The messages are encrypted using session keys.
+ */
+
+#pragma once
+
+#include <messaging/ExchangeTransport.h>
+#include <transport/TransportMgr.h>
+
+namespace chip {
+namespace Transport {
+class BLE;
+}
+
+class SessionEstablishmentTransport : public Messaging::ExchangeTransport
+{
+public:
+    SessionEstablishmentTransport() {}
+
+    virtual ~SessionEstablishmentTransport() {}
+
+    CHIP_ERROR Init(Transport::BLE * transport, TransportMgrBase * transportMgr)
+    {
+        ReturnErrorCodeIf(transport == nullptr && transportMgr == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+        mBLETransport = transport;
+        mTransportMgr = transportMgr;
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR OnMessageReceived(uint16_t protocol, uint8_t type, const Transport::PeerAddress & peerAddress,
+                                 Messaging::ReliableMessageContext & reliableMessageContext,
+                                 MessageReliabilityInfo & reliabilityInfo) override;
+
+    const Transport::PeerAddress & GetPeerAddress() const { return mPeerAddress; }
+
+    void SetPeerAddress(const Transport::PeerAddress & address) { mPeerAddress = address; }
+
+protected:
+    CHIP_ERROR SendMessageImpl(SecureSessionHandle session, PayloadHeader & payloadHeader, System::PacketBufferHandle && message,
+                               EncryptedPacketBufferHandle * retainedMessage) override;
+
+    bool MessagePermitted(uint16_t protocol, uint8_t type) override;
+
+    bool IsTransportReliable() override { return (mBLETransport != nullptr); }
```

Comment:
UDP transport is also not reliable, I would rename it to IsBleTransport

Reply:
The base class (ExchangeTransport) provides this virtual function. The same function will be overridden for TCP based CASE session establishment. So I think it's better to keep the current name.

Reply:
My concern here is UDP is also not a reliable Transport, why only check BLE? We should use CRMP for both all unreliable  transports.

Reply:
It'll return false if UDP is used as the transport. So CRMP will kick in for the UDP transport. For BLE transport, it'll return true.

Reply:
Updated the code for more clarity.